### PR TITLE
[aarch64] rework gemm convolution and matmul operators to support caching

### PR DESCRIPTION
**Pull request by snadampal**
Rework the gemm convolution and matmul operators' state machine for acl object creation and configure to take advantage of onednn primitive caching feature.

# Description
Rework the gemm convolution and matmul operators' state machine for acl object creation and configure to take advantage of onednn primitive caching feature.

Fixes # (github issue)

# Checklist

## General

- [ x] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
make test: all passed except test#79 ad test#98. they are failing on master (without this commit) as well.
- [x ] Have you formatted the code using clang-format?

## Performance improvements

- [ ] Have you submitted performance data that demonstrates performance improvements?

### New features

- [ ] Have you published an RFC for the new feature?
- [ ] Was the RFC approved?
- [ ] Have you added relevant tests?

### Bug fixes

- [ ] Have you included information on how to reproduce the issue (either in a github issue or in this PR)?
- [ ] Have you added relevant regression tests?

## [RFC](https://github.com/oneapi-src/oneDNN/tree/rfcs) PR

- [ ] Does RFC document follow the [template](https://github.com/oneapi-src/oneDNN/blob/rfcs/rfcs/template.md#onednn-design-document-rfc)?
- [ ] Have you added a link to the rendered document?

### CODE REVIEW

**Comment by lilh9598** [comment]
Look great! In my opinion, It will be helpful for the performance of TF-oneDNN-ACL when inter_op_parallelism_threads is not equal to 1, but may cause more memory consumption.

File: src/cpu/aarch64/matmul/acl_matmul.hpp [hunk 1]
```diff
@@ -96,12 +96,24 @@ struct acl_matmul_t : public primitive_t {
             acl_obj_->src_acc_tensor.allocator()->init(amp.src_acc_info);
             acl_obj_->transA.configure(
                     &acl_obj_->src_acc_tensor, &acl_obj_->src_tensor);
+
+            acl_obj_->src_intermediate_tensor.allocator()->init(amp.src_info);
+            acl_obj_->src_intermediate_tensor.allocator()->allocate();
```
Comment (mgouicem) [inline src/cpu/aarch64/matmul/acl_matmul.hpp:101]: Unfortunately, this is contrary to oneDNN memory model. oneDNN typically does not hog memory inside primitive objects.
oneDNN calls these temporary memory "scratchpad memory", and it is either allocated during execution (library mode) or provided by user (user mode). See [this page ](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_scratchpad.html#primitive-attributes-scratchpad) for more details.

**Review by mgouicem — COMMENTED** [review]
Thanks for the PR. The primitive cache part is interesting. Tagging @jondea @cfRod for more comments.

However, the memory allocation inside primitive is contrary to oneDNN scratchpad management policies/controls.

Also, please modify the commit naming to comply with oneDNN commit names format specified in [CONTRIBUTING.md](https://github.com/oneapi-src/oneDNN/blob/master/CONTRIBUTING.md#code-contribution-guidelines)

File: src/cpu/aarch64/matmul/acl_matmul.hpp (hunk 1 above)
Comment (snadampal) [reply src/cpu/aarch64/matmul/acl_matmul.hpp:101]: Hi @mgouicem, thanks for the feedback. I have some high level question on the onednn primitive caching feature, could you please clarify:
if we expect the primitive caching to not to hold on to any primitive memory, then what exactly the caching feature is saving in the execution time? In the entire primitive create/configure/execute states I see tensor memory allocation is the major time consuming operation.  Given tensors could be of > 100MB for some CNN networks, allocating the buffers during execute, every time, may not be acceptable. I consider these intermediate tensors as 'workspace' buffers as these are not known to user to be able to provide them via scratchpad.

Comment (mgouicem) [reply src/cpu/aarch64/matmul/acl_matmul.hpp:101]: Hi @snadampal , the primitive cache mainly reduces overheads of creating primitive objects. For the acl implementations, it might not be much, but for other implementations (e.g. sve, and gpu), this overhead of generating executable binaries taylored to a given problem shapes can be large.

Regarding memory allocation, oneDNN is typically called within a larger FWK (e.g. tensorflow, pytorch, OpenVINO, ...). Those have dedicated memory allocator that allow to get memory pages from the OS, and just reuse them across multiple calls of oneDNN. To do that, they create primitives using the scratchpad user_mode, and pass the scratchpad memory directly to the primitive execution function. Hence they manage the lifetime of the scratchpad as they wish.

If we were to keep scratchpad memory alive inside the oneDNN primitive cache, this could lead to large memory consumption from oneDNN, and this is not acceptable on some devices (e.g. client devices, GPUs with limited memory, ...).

Hope this helps clarify this design choice.

Comment (snadampal) [reply src/cpu/aarch64/matmul/acl_matmul.hpp:101]: @mgouicem really appreciate your time for putting all these details, I got the rational for the current design, thank you!
Now I understand why you have primitive and resource objects separate and why the resources are not cached. Yes, both TensorFlow and PyTorch are using the scratchpad user_mode, but for aarch64 I see ACL managing the memory allocations and these tensors are being imported everywhere else. I will revisit the tensor buffer management flow again.

File: src/cpu/aarch64/acl_gemm_convolution.hpp [hunk 2]
```diff
@@ -144,10 +139,9 @@ struct acl_gemm_convolution_fwd_t : public primitive_t {
     }
 
 private:
-    // To guard the const execute_forward(), the mutex must be 'mutable'
-    mutable std::mutex mtx;
     status_t execute_forward(const exec_ctx_t &ctx) const;
     const pd_t *pd() const { return (const pd_t *)primitive_t::pd().get(); }
+    std::unique_ptr<acl_obj_t<arm_compute::NEGEMMConvolutionLayer>> acl_obj_;
```
Comment (densamoilov) [inline src/cpu/aarch64/acl_gemm_convolution.hpp:144]: Hi @snadampal, as far as I remember the main reason of putting the ACL objects into resources was to avoid storing them in the `primitive_t` because those ACL objects are not immutable. All the entities that reside in `primitive_t` must be immutable due to primitive cache design, otherwise there might be problems with concurrent execution of identical primitives that were fetched from the primitive cache.

Per my understanding the ACL object is still not immutable therefore we have to keep it in the resource. Am I missing something?

Comment (snadampal) [reply src/cpu/aarch64/acl_gemm_convolution.hpp:144]: Hi @densamoilov , could you please elaborate on this? What do you mean by identical primitives here?

>  otherwise there might be problems with concurrent execution of identical primitives that were fetched from the primitive cache.

I have verified these changes with PyTorch. However, with TensorFlow I see out of order execution and mismatch for the src and weight tensor pair randomly. I'm suspecting it due to concurrent execution of the cached primitives, and debugging it right now.
Btw, these changes provide significant performance improvement for CNNs with large number of channels, hence I would like to better understand the concurrent execution constraint and explore options to handle them.

Comment (nSircombe) [reply src/cpu/aarch64/acl_gemm_convolution.hpp:144]: > Per my understanding the ACL object is still not immutable therefore we have to keep it in the resource. Am I missing something?

Yep, this was the original motivation back when we first discussed ACL integration - nothing has changed on the ACL side since then.

Comment (densamoilov) [reply src/cpu/aarch64/acl_gemm_convolution.hpp:144]: > I would like to better understand the concurrent execution constraint and explore options to handle them.

The primitive cache is a global object that stores created primitive implementations with `shared_ptr`. A primitive implementation is a class derived from `primitive_t`. There is a difference between primitive and primitive implementation. A primitive is a class named `primitive_iface_t` that includes the primitive implementation (`primitive_t`), scratchpad and resource. The `primitive_iface_t` is created every time when a user creates a primitive. In fact, a pointer to `primitive_iface_t` is what the library returns to the user. Since primitive implementations can be shared between threads we have to make sure that they are immutable to avoid race condition.

It can easily happen that a user creates and executes two identical primitives in different threads.
```cpp
#pragma omp parallel num_threads(2)
{
    // While the convolution handles are different they both point to the
    // same primitive implementation internally.
    convolution_forward conv({...});
    // If the ACL object is stored in the primitive implementation then this will
    // lead to a race condition.
    conv.execute(...);
}
```
In order to address such issues we introduced an abstraction called resource. The resource is a storage that is unique for each created primitive and is used to store data that cannot be shared.

Comment (snadampal) [reply src/cpu/aarch64/acl_gemm_convolution.hpp:144]: Thanks for the details,@densamoilov. Yes, I have seen this note for primitive iface class 
https://github.com/oneapi-src/oneDNN/blob/v2.6/src/common/primitive.hpp#L270
that's when I started moving the acl object from resource to primitive so that it will get cached.

And from the below note for the resource class, I didn't infer acl objects be part of this cache exception list.
https://github.com/oneapi-src/oneDNN/blob/v2.6/src/common/primitive.hpp#L189

thanks for the code snippet above, it makes things more clear. So, for identical primitives, 
1. there will be a single primitive implementation object but having multiple resources (and hence the multiple acl object) associated with it, right? 
2. they are differentiated through the resource mapper object set in each execution context, right?

I'm curious how it is handled for other engines (eg. gpu) where resource allocation will be an overhead every time. and I didn't see the "resource_t" implementation for cpu/x86 engine. How is it handled there?

Comment (mgouicem) [reply src/cpu/aarch64/acl_gemm_convolution.hpp:144]: Just a note: it seems that the acl object have to reside in ressource. However, the allocated memory does not have to.
If you rely on scratchpad for memory allocation (how it should be, given how they are used), the memory allocation would no more happen upon each execution. At execution time, the pointers in scratchpad would be imported to the acl tensor objects. Does this make sense?

A pre-requisite though would be for acl tensors to have a method to query size (in order to know the size of scratchpad).

Comment (snadampal) [reply src/cpu/aarch64/acl_gemm_convolution.hpp:144]: Yes, looks like scratchpad or acl memory pool is the option I see to avoid redundant memory allocations. Thanks @mgouicem.

Comment (densamoilov) [reply src/cpu/aarch64/acl_gemm_convolution.hpp:144]: > 1. there will be a single primitive implementation object but having multiple resources (and hence the multiple acl object) associated with it, right?

Yes. 
> 2. they are differentiated through the resource mapper object set in each execution context, right?

Not exactly. I simplified my explanation when I said there is 1 resource instance per primitive. In practice it's possible that a primitive implementation includes another primitive implementation and that primitive implementation includes another one and so on. We call them nested primitive implementations. In the case of ACL each of them may have their own ACL object and therefore their own resource instance. This requires a mechanism whereby the nested primitive implementations can obtain their resource instances. The mechanism is implemented via the resource mapper abstraction, where key is address of a primitive implementation and value is the corresponding resource instance.
To sum it up:
* There is always only 1 instance of resource mapper **per primitive**
* In the case when **there are no** nested primitive implementations the resource mapper contains only 1 resource instance for the main primitive implementation
* In the case when **there are** nested primitive implementations the resource mapper contains resource instance for the main primitive implementation and all nested ones

If we want to cache ACL objects then we either need to be able to clone them (similar to OpenCL kernels) at execution time or make sure that they are immutable. I don't see other ways to cache them.

> I didn't see the "resource_t" implementation for cpu/x86 engine. How is it handled there?

That's because for CPU we have JIT kernels, which are represented as immutable arrays of bytes therefore we can cache them.

Comment (snadampal) [reply src/cpu/aarch64/acl_gemm_convolution.hpp:144]: @densamoilov thanks! I'm exploring few options with ACL memory manager level caching the allocations outside the acl object. Will update this PR once the changes are ready for review.

**Comment by snadampal** [comment]
I will ping once the ACL objects are made immutable.

**Comment by snadampal** [comment]
will reopen once ACL objects are made immutable.
